### DISCUSSION TITLE
bepuphysics in Godot

### DISCUSSION THREAD

**Discussion by AdrianPokojny** [bepu/bepuphysics2#294] [General]
I made a quick test to check the performance of Godot using bepuphysics, it looks quite promising, here's a video of the ColosseumDemo in Godot 

https://www.youtube.com/watch?v=cmoGWT5fOAc

Performance aren't as amazing as it should be but it's way better than the Unity integration, I'll post updates if I manage to make something nice out of it

**Comment by RossNordby** [comment]
Nice!

Curiously, I may have done almost exactly the same test a couple of weeks ago :P
![image](https://github.com/bepu/bepuphysics2/assets/657209/3ad7c958-ba97-4522-927a-c773fefed151)

I also noticed it was >5x slower than I expected, which I'm a bit confused by. My understanding was that godot's C# integration used CoreCLR and the runtime-relevant information I checked seemed to support that. There's a cost to jumping through the interop layers, but the slowdown persisted in regions that should have had no interop whatsoever. I didn't have time to do a full investigation, unfortunately.

If you manage to figure out what's going on with that, I'd be interested in hearing about it!

**Comment by AdrianPokojny** [comment]
Such a funny coincidence that we made the same experiment just a few weeks aparts :D

Yes sadly it's slower, someone told me building the game should cancel most of the performance penalty for C# in Godot but as shown in the video it's still not as efficient. 

My original idea was to make the transform in C# and transfer the variable in GDscript to apply the transform on the GDscript side as the C# API is slower when it calls godot methods. It has indeed a slight impact but not worth it. Maybe I should find a way to not use Godot at all on the C# side and just get variables through GDscript cos maybe what slowdown the simulation is the conversion of System.Numerics.Vector3 to Godot.Vector3 and such things.

As you said there's also a cost jumping through interop layers.

It made me wanna try the same experiment on Stride Engine as it's mainly written in C# and see the results !

**Comment by AdrianPokojny** [comment]
https://www.reddit.com/r/godot/comments/1725cyf/comment/k3yq6p8/?utm_source=share&utm_medium=web2x&context=3### 

![Capture](https://github.com/bepu/bepuphysics2/assets/143230516/1fc22d8b-1de0-4739-baf0-280b41373592)

Interesting, I don't think it will solve the problem, but who knows

**Reply by AdrianPokojny** [reply]
https://github.com/godotengine/godot-proposals/issues/7895

Ok finally it might solve the problem, I'll try again when the update will be released

**Reply by jonri** [reply]
The other interesting thing about C# moving to GDExtension is you should be able to subclass PhysicsServer3DExtension (https://docs.godotengine.org/en/stable/classes/class_physicsserver3dextension.html) which at least used to be problematic to attempt from C# before.  This lets you register the physics server as a true replacement for the default one, similar to the way the Jolt extension works.  Maybe it's already doable and you've looked into it, but if not this might help avoid some of the manual transforms that it sounded like you were doing currently.

**Comment by lucasthomaz97** [comment]
I discovered this physics engine yesterday and desired to use it with Godot on my own games, today  i found this discussion, now that Godot 4.3 is the current version, the improvement happened? How is it going?

**Reply by RossNordby** [reply]
Dunno! I'm perpetually low on time these days, so if you wanted to poke around and try running some benchmarks that'd be neat.
